File: Sum_of_Subarray_Minimums.cpp
class Solution {
public:
    int sumSubarrayMins(vector<int>& arr) {
        stack<vector<int>> stack;
        vector<int> res(arr.size(), 0);

        for (int i = 0; i < arr.size(); i++) {
            // use ">=" to deal with duplicate elements
            while (!stack.empty() && stack.top()[0] >= arr[i]) {
                stack.pop();
            }

            int j = stack.empty() ? -1 : stack.top()[1];
            res[i] = stack.empty() ? arr[i] * (i + 1) : res[j] + arr[i] * (i - j);
            stack.push({arr[i], i});
        }

        const long MOD = static_cast<long>(1e9 + 7);
        long ans = 0;

        for (int i = 0; i < arr.size(); i++) {
            ans = (ans + static_cast<long>(res[i])) % MOD;
        }
        return static_cast<int>(ans);        
    }
};

File: Sum_of_Subarray_Minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        Stack<int[]> stack = new Stack<>();
        int[] res = new int[arr.length];
        
        for (int i = 0; i < arr.length; i++) {
            // use ">=" to deal with duplicate elements
            while (!stack.isEmpty() && stack.peek()[0] >= arr[i]) {
                stack.pop();
            }

            int j = stack.isEmpty() ? -1 : stack.peek()[1];
            res[i] = stack.isEmpty() ? arr[i] * (i + 1) : res[j] + arr[i] * (i - j);
            stack.push(new int[]{arr[i], i});
        }

        long MOD = (long) 1000000007;
        long ans = 0;

        for (int i = 0; i < arr.length; i++) {
            ans = (ans + (long) res[i]) % MOD;
        }
        return (int) ans;        
    }
}
